rewrite abstract and introduction (#24)
danwing/cidfi: draft-wing-cidfi.md

--- abstract

> Host to Network signaling and Network to Host signaling can improve

Reviewer: ```suggestion
Host-to-network signaling and network-to-host signaling can improve
```

> the user experience to meet network's constraints and to provide

Reviewer: ```suggestion
the user experience to adapt to network's constraints and share expected application needs, and thus to provide
```

differentiated service to individual packets.

This document describes how clients can communicate with their nearby
network elements so their QUIC and DTLS streams can be augmented with
information about network conditions and packet importance.  With
optional server support individual packets can receive differentiated

> service.

Reviewer: ```suggestion
service. The proposed approach covers both directions of a flow.
```

content (e.g., improves video quality which consumes more bandwidth which
then gets dropped by the network).

Due to network constraints a network element will need to discard a packet
or prioritize a packet ahead of other packets within a single UDP 4-tuple.  The network element
cannot make an informed decision of which packet to discard or pirioritize without
knowing the relative

> importance of the packet.  Metadata carried in each packet can inform the

Reviewer: ```suggestion
importance of the packet.  Metadata carried in each packet (or a subset) can inform the
```

network element
to improve the user experience.

This document defines CIDFI (pronounced "sid fye") which provides
host-to-network (h2n) and network-to-host (n2h) signaling.  CIDFI
always requires two components:  a CIDFI-aware client with a QUIC or

> DTLS stack client QUIC and at least one CIDFI-aware network element

Reviewer: ```suggestion
DTLS client stack and at least one CIDFI-aware Network Element
```

(CNE).  The client sends metadata about incoming CIDs to the CNE (h2n
signaling) and the CNE sends metadata about its view of the network
(n2h signaling).  The CNE-aware client can inform the server of
network constraints it learned via n2h signaling such as by changing
codecs (e.g., using offer/answer {{?RFC3264}}) or requesting a lower-
quality streaming video.

> Optionally, a CNE-aware server can be deployed which then allows

Reviewer: ```suggestion
Optionally, a CIDFI-aware server can be deployed which then allows
```

> differentiating classes of packets within a flow.  Those

Reviewer: ```suggestion
differentiating classes of packets within a flow (e.g., specific QUIC streams).  Those
```

differentiated packets can be client-to-server or server-to-client.

> By including a CNE-aware server in the system the CNE can provide

Reviewer: ```suggestion
By including a CIDFI-aware server in the system the CNE can provide
```

The CIDFI-aware client establishes a TLS connection with the
CIDFI-aware network elements (Wi-Fi access point, edge router, and RAN
router in the above diagram).  Over this connection it receives
network performance information (n2h) and it sends mapping of (QUIC or DTLS)
Destination CIDs to packet importance (h2n).

The design creates new state in the CIDFI-aware network elements for

> mapping from Destination CID to the packet metadata and maintaining

Reviewer: ```suggestion
mapping from Destination CIDs to the packet metadata and maintaining
```

triggers to update the client if the network characteristics change,

> and to maintain a TLS channel with the client.

Reviewer: ```suggestion
and to maintain a (D)TLS channel with the client.
```

Client Authorization:
: The client authorizes each CIDFI-aware network element (CNE) to participate in CIDFI
for each QUIC (or DTLS) flow.

Same Server:
: When the server also participates in CIDFI, the same QUIC connection is used for CIDFI
communication with that server,
which ensures it arrives at the same server instance even in the presence of
network translators (NAT) or server-side ECMP load balancers or server-side CID-aware
load balancers {{?I-D.ietf-quic-load-balancers}}.

Privacy:
: The host-to-network signaling of the mapping from packet metadata to CID is only sent to CIDFI-aware network
elements (CNEs) and is protected by TLS.  The network-to-host signaling of network metadata is protected by TLS.  For

> CIDFI to operate, the CNE never needs the server's identity, and the CNE is never provided decryption keys for

Reviewer: ```suggestion
CIDFI to operate, a CNE never needs the server's identity, and a CNE is never provided decryption keys for
```

When a QUIC client (or DTLS-CID) client connects to a QUIC (or DTLS-CID) server, the client:

  1. learns if the server supports CIDFI
     and obtains its mapping of transmitted Destination CIDs to metadata, described
     in {{server-supports-cidfi}}.
  2. proves ownership of its UDP 4-tuple to
     the on-path CNEs, described in {{ownership}}.
  3. performs initial metadata exchange
     with the CIDFI network element and server, and server and network element,
     described in {{initial-metadata-exchange}}.
  4. for the duration of the connection, receives network-to-host and performs
     host-to-network updates as network conditions or network requirements change,

>      described in {{ongoing}}.

Reviewer: ```suggestion
     described in {{ongoing}}. Some policies are provided to CNEs to control which network changes can triggers updating clients.
```

On initial connection to a QUIC server, the client includes a new QUIC
transport parameter CIDFI ({{iana-tp}}) which is remembered for 0-RTT.

If the server does not indicate CIDFI support, the client can still
perform CIDFI -- but does not expect different CIDs to indicate

> differentiated behavior.  The client can still signal its CNE about

Reviewer: ```suggestion
differentiated behavior.  The client can still signal its CNE(s) about
```

# Ongoing Signaling {#ongoing}

Throughout the life of the connection host-to-network and network-to-host

> signaling is updated whenever charactertics change.

Reviewer: ```suggestion
signaling is updated whenever characteristics change. Still, some policies are provided to control when these updates are triggers. Such policies are meant to preserve the connection stability.
```

# Interaction with Load Balancers {#load-balancers}

> QUIC servers are frequently behind CID-aware load balancers {{?I-D.ietf-quic-load-balancers}}.

Reviewer: ```suggestion
QUIC servers are likely to be behind CID-aware load balancers {{?I-D.ietf-quic-load-balancers}}.
```
